### PR TITLE
enable vcpkg for all platforms on CI

### PR DESCRIPTION
**Pull request by Arcnor**
## Pull Request Type

- [x] GitHub Workflow changes
- [ ] Documentation or Wiki changes
- [x] Build and Dependency changes
- [ ] Runtime changes
  - [ ] Render changes
  - [ ] Audio changes
  - [ ] Input changes
  - [ ] Network changes
  - [ ] Other changes

### Description
Enable vcpkg on all platforms, include & link SDL2 for all of them as well

### Related Issues
Relates to #241 

### Checklist

- [x] I have tested my changes locally and verified that they work as intended.
- [x] I have documented any new or modified functionality.
- [x] I have reviewed the changes to ensure they do not introduce any unnecessary complexity or duplicate code.
- [x] I understand that by submitting this pull request, I am agreeing to license my contributions under the project's license.

### CODE REVIEW

File: Descent3/CMakeLists.txt
Comment (Lgt2x) [inline Descent3/CMakeLists.txt:303]: trying to understand this line. Can't we assume we always have the SDL2::SDL2 target ? How would the find package only find the static build ?

File: vcpkg.json
Comment (Lgt2x) [inline vcpkg.json:8]: 2.30.2 has not been released on VCPKG yet ?

**Review by Lgt2x — COMMENTED** [review]
Update README instructions to use VCPKG on Linux ? Good job otherwise

**Review by winterheart — APPROVED** [review]
I don't mind against using vcpkg on Linux and macOS, but most likely Linux distributions will force using system libraries like SDL2 or spdlog from their repositories, so we must provide them ability to do so (again, proposed changes allows that, so OK from me).

File: .github/workflows/build.yml
Comment (Arcnor) [reply .github/workflows/build.yml:71]: We do not, but if I remove them the current build will fail until we actually remove SDL1 from the source :)

**Comment by Arcnor** [comment]
> Update README instructions to use VCPKG on Linux ? Good job otherwise

Was hoping you could help me with that :D. I had a change where I basically added the same line as on Windows, but it seems macOS and Linux have very specific instructions, so in the same vein maybe we can provide exact instructions. If you can try to install vcpkg and tell me the proper paths I can add them.

File: vcpkg.json
Comment (Arcnor) [reply vcpkg.json:8]: No, it has not.

File: Descent3/CMakeLists.txt
Comment (Arcnor) [reply Descent3/CMakeLists.txt:303]: If using the vcpkg target, there is always either a `SDL2::SDL2` target for the dynamic version, or a `SDL2::SDL2-static` one. The `find_package` will find the associated CMake package, but a package can have multiple targets, so this line basically just tries to find first the dynamic one using generator expressions, then the static one.

**Comment by Arcnor** [comment]
> I don't mind against using vcpkg on Linux and macOS, but most likely Linux distributions will force using system libraries like SDL2 or spdlog from their repositories, so we must provide them ability to do so (again, proposed changes allows that, so OK from me).

Well, this is just for building, I think people can use whatever they want for running, but specifying what they need for building is something I think we can do, and in this case it can be "vcpkg is required". That said, I can wait until Ryan's PR is merged, then I add all of this on top of his stuff, and basically fallback to the system one if it cannot find the `CONFIG` one.

File: vcpkg.json
Comment (Arcnor) [reply vcpkg.json:9]: It's good practice IMHO, as this is what we expect as a minimum so get it "documented" is nice. Also if you have an older vcpkg ports baseline and you suddenly install `2.28.0` instead for example there will be no error (that said we specify the baseline here as well, so this particular issue won't happen).

File: .github/workflows/build.yml
Comment (Lgt2x) [reply .github/workflows/build.yml]: Please mark those concerned for deletion (ie on a separate line with a comment)

File: Descent3/CMakeLists.txt
Comment (Lgt2x) [reply Descent3/CMakeLists.txt:303]: got it

**Comment by Lgt2x** [comment]
> > Update README instructions to use VCPKG on Linux ? Good job otherwise
> 
> Was hoping you could help me with that :D. I had a change where I basically added the same line as on Windows, but it seems macOS and Linux have very specific instructions, so in the same vein maybe we can provide exact instructions. If you can try to install vcpkg and tell me the proper paths I can add them.

On a fresh ubuntu:latest docker image, I can build using the following commands (first line is new VCPKG requirements):
```
apt install -y git cmake g++ curl zip unzip tar ninja-build pkg-config python3 python3-distutils-extra python3-jinja2 autoconf autopoint libtool build-essential gperf 
apt install -y --no-install-recommends ninja-build cmake g++ libsdl1.2-dev libsdl-image1.2-dev libncurses-dev libxext6:i386 zlib1g-dev spdlog-dev
git clone https://github.com/microsoft/vcpkg
./vcpkg/bootstrap-vcpkg.sh
VCPKG_ROOT=./vcpkg cmake --preset linux
cmake --build --preset linux --config Debug```

However, I'm not sure about recommending this approach to anyone wanting to build on Linux. It has much more requirements in order to bulid the VCPKG packages, and the build is way longer. In addition, VCPKG errors can be obscure and harder to debug (it's not just written plainly "Missing package X")

**Comment by JeodC** [comment]
I have added this along with github caching on branch `cache`, and ran a workflow dispatch here: https://github.com/DescentDevelopers/Descent3/actions/runs/8929731898

Please inspect and modify as needed.

**Comment by JeodC** [comment]
If we merge `cache` branch into main, we can knock out the following while doing so: #257, #242, #276, #258

**Comment by Arcnor** [comment]
> > > Update README instructions to use VCPKG on Linux ? Good job otherwise
> > 
> > Was hoping you could help me with that :D. I had a change where I basically added the same line as on Windows, but it seems macOS and Linux have very specific instructions, so in the same vein maybe we can provide exact instructions. If you can try to install vcpkg and tell me the proper paths I can add them.
> 
> On a fresh ubuntu:latest docker image, I can build using the following commands (first line is new VCPKG requirements):
> ```
> apt install -y git cmake g++ curl zip unzip tar ninja-build pkg-config python3 python3-distutils-extra python3-jinja2 autoconf autopoint libtool build-essential gperf 
> apt install -y --no-install-recommends ninja-build cmake g++ libsdl1.2-dev libsdl-image1.2-dev libncurses-dev libxext6:i386 zlib1g-dev spdlog-dev
> git clone https://github.com/microsoft/vcpkg
> ./vcpkg/bootstrap-vcpkg.sh
> VCPKG_ROOT=./vcpkg cmake --preset linux
> cmake --build --preset linux --config Debug
> ```
> 
> However, I'm not sure about recommending this approach to anyone wanting to build on Linux. It has much more requirements in order to bulid the VCPKG packages, and the build is way longer. In addition, VCPKG errors can be obscure and harder to debug (it's not just written plainly "Missing package X")

To be honest, those packages you'll need anyway for any kind of development, and the build is longer only the very first time, which is not a useful metric when we have cache IMHO. On the CI agents I had to add none, but I'm not sure that's a useful metric either.

About the errors, I'll need an example to know why you think they are hard to understand, but vcpkg only calls CMake under the hood, and it clearly states where the log with the errors is.

Finally, if you think this is not useful let's just close and maybe remove `vcpkg` itself like I mentioned on Discord, we just need consensus on that.

**Comment by JeodC** [comment]
Superseded by #284

**Comment by Arcnor** [comment]
@JeodC what you just did makes no sense. You don't "superseed" by copying manually these changes into new commits you just create, you create your changes on top (or rebase) if you're going to add them. I've reopened this and closed yours.

**Comment by Jayman2000** [comment]
> I've reopened this and closed yours.

A similar thing happened to #281. Do you think that that PR should be reopened? For context, there’s two PRs that would fix #276, but they’re both closed at the moment.

**Comment by JeodC** [comment]
> > I've reopened this and closed yours.
> 
> A similar thing happened to #281. Do you think that that PR should be reopened? For context, there’s two PRs that would fix #276, but they’re both closed at the moment.

I did not expect #284 to be abruptly closed as it was. I'll go ahead and reopen it.

**Comment by Jayman2000** [comment]
> I tried to reopen it but the branch it originated from was force-pushed or recreated so the button is blank.

OK. [I created a new PR that uses the same branch.](https://github.com/DescentDevelopers/Descent3/pull/287)

**Comment by Lgt2x** [comment]
> > > > Update README instructions to use VCPKG on Linux ? Good job otherwise
> > > 
> > > 
> > > Was hoping you could help me with that :D. I had a change where I basically added the same line as on Windows, but it seems macOS and Linux have very specific instructions, so in the same vein maybe we can provide exact instructions. If you can try to install vcpkg and tell me the proper paths I can add them.
> > 
> > 
> > On a fresh ubuntu:latest docker image, I can build using the following commands (first line is new VCPKG requirements):
> > ```
> > apt install -y git cmake g++ curl zip unzip tar ninja-build pkg-config python3 python3-distutils-extra python3-jinja2 autoconf autopoint libtool build-essential gperf 
> > apt install -y --no-install-recommends ninja-build cmake g++ libsdl1.2-dev libsdl-image1.2-dev libncurses-dev libxext6:i386 zlib1g-dev spdlog-dev
> > git clone https://github.com/microsoft/vcpkg
> > ./vcpkg/bootstrap-vcpkg.sh
> > VCPKG_ROOT=./vcpkg cmake --preset linux
> > cmake --build --preset linux --config Debug
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > However, I'm not sure about recommending this approach to anyone wanting to build on Linux. It has much more requirements in order to bulid the VCPKG packages, and the build is way longer. In addition, VCPKG errors can be obscure and harder to debug (it's not just written plainly "Missing package X")
> 
> To be honest, those packages you'll need anyway for any kind of development, and the build is longer only the very first time, which is not a useful metric when we have cache IMHO. On the CI agents I had to add none, but I'm not sure that's a useful metric either.

True, only the first build is slower.

> 
> About the errors, I'll need an example to know why you think they are hard to understand, but vcpkg only calls CMake under the hood, and it clearly states where the log with the errors is.

I lost my terminal session since then, but the 2 layers VCPKG, and CMake under the hood did help obscure the real error. 

> 
> Finally, if you think this is not useful let's just close and maybe remove `vcpkg` itself like I mentioned on Discord, we just need consensus on that.

Maybe it's just me. I think we can continue with these changes to the CI, it's an important step towards SDL2. We'll just need a rebase and we can merge

**Review by Lgt2x — APPROVED** [review]
Ok to merge when README instructions are updated for Linux (maybe mac?) using the new package requirements for VCPKG builds (see my message above)

File: .github/workflows/build.yml
Comment (Lgt2x) [inline .github/workflows/build.yml:70]: I can't remember if I asked before but libsdl2-dev, zlib & spdlog should now be provided by VCPKG ?

**Review by bryanperris — APPROVED** [review]

**Comment by Lgt2x** [comment]
this PR needs a rework with the removal of spdlog.

**Comment by Lgt2x** [comment]
Closing as inactive and outdated, many 3rd party changes since this was first open. Still, it's an idea I'd like to see implemented

**Pull request closed without merging**
